### PR TITLE
Feature/tls for mysql

### PR DESCRIPTION
**Pull request by daniellowtw**

### CODE REVIEW

**Comment by daniellowtw** [comment]
@mwitkow PTAL

**Comment by daniellowtw** [comment]
@jonny-improbable

File: conf/defaults.ini [hunk 1]
```diff
@@ -62,6 +62,13 @@ password =
 # For "postgres" only, either "disable", "require" or "verify-full"
 ssl_mode = disable
 
+# For "mysql" only, either true, false, skip-verify
+tls_mode = false
```
Comment (mwitkow) [inline conf/defaults.ini]: can you reuse the `ssl_mode` for this?

File: conf/defaults.ini [hunk 2]
```diff
@@ -62,6 +62,13 @@ password =
...
+ca_path =/etc/mysql/ca.pem
+client_key_path =/etc/mysql/client-key.pem
+client_cert_path =/etc/mysql/client-cert.pem
+server_name =localhost
```
Comment (mwitkow) [inline conf/defaults.ini]: name this something like `server_cert_name`

**Comment by mwitkow** [comment]
Please update docs/sources/installation/configuration.md to document the changes you made

File: pkg/services/sqlstore/sqlstore.go [hunk 3]
```diff
@@ -115,6 +126,15 @@ func getEngine() (*xorm.Engine, error) {
 	case "mysql":
 		cnnstr = fmt.Sprintf("%s:%s@tcp(%s)/%s?charset=utf8",
 			DbCfg.User, DbCfg.Pwd, DbCfg.Host, DbCfg.Name)
+		if mysqlConfig.UseTls != "false" {
+			fmt.Printf("%+v", mysqlConfig)
```
Comment (mwitkow) [inline pkg/services/sqlstore/sqlstore.go]: remove the printf

File: pkg/services/sqlstore/tls_mysql.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package sqlstore
+import (
+	"crypto/tls"
```
Comment (mwitkow) [inline pkg/services/sqlstore/tls_mysql.go:4]: you should be using tabs for go, right?

File: pkg/services/sqlstore/tls_mysql.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	"crypto/x509"
+	"io/ioutil"
+)
+
+func makeCert(tlsPoolName string, config MySQLConfig) (*tls.Config, error) {
+	rootCertPool := x509.NewCertPool()
+	pem, err := ioutil.ReadFile(config.CaPath)
+	if err != nil {
+		return nil, err
```
Comment (mwitkow) [inline pkg/services/sqlstore/tls_mysql.go]: improve error reporting throughout this code:

`fmt.Errorf("could not read DB CA Cert path: %v", err)`

File: pkg/services/sqlstore/tls_mysql.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	clientCert = append(clientCert, certs)
+	tlsConfig := &tls.Config{
+		RootCAs:      rootCertPool,
+		Certificates: clientCert,
+	}
+	if config.ServerName != "" {
+		tlsConfig.ServerName = config.ServerName
+	}
+
+	if config.UseTls == "skip-verify" {
```
Comment (mwitkow) [inline pkg/services/sqlstore/tls_mysql.go]: make this map to an existing `ssl_mode`

**Comment by mwitkow** [comment]
ROFL

File: conf/defaults.ini (hunk 1 above)
Comment (daniellowtw) [inline conf/defaults.ini]: The dsn structure for having ssl is different for postgres and mysql. They accept different values. I could make it look similar by doing mysql_ssl_mode

File: pkg/services/sqlstore/tls_mysql.go (hunk 4 above)
Comment (daniellowtw) [inline pkg/services/sqlstore/tls_mysql.go:4]: It is using tabs

**Comment by daniellowtw** [comment]
PTAL

**Pull request merged**